curta/src/chip/hash/blake/blake2b/mod.rs

```diff
@@ -0,0 +1,1177 @@
+            &t,
+            &last_chunk_bit,
+            &cycle_12_start_bit,
+            &cycle_12_end_bit,
+            &unused_row,
+            operations,
+        );
+
+        for ((h_in, init), h_out) in h_input.iter().zip(initial_hash.iter()).zip(h_output.iter()) {
+            self.set_to_expression_transition(
+                &h_in.next(),
+                last_chunk_bit.expr()
+                    * (cycle_12_end_bit.expr() * init.expr()
+                        + cycle_12_end_bit.not_expr() * h_out.expr())
+                    + (last_chunk_bit.not_expr() * h_out.expr()),
+            );
+        }
+
+        self.add_bus_constraints(
+            clk,
+            bus,
+            bus_channel_idx,
+            num_chunks,
+            &cycle_12_start_bit,
+            &cycle_12_end_bit,
+            &m,
+            &t,
+            &last_chunk_bit,
+            &h_output,
+            &padding_bit,
+            &msg_chunks,
+            &t_public,
+            &last_chunk_bit_public,
+            &hash_state,
+        );
+
+        BLAKE2BGadget {
+            padding_bit,
+            cycle_12_start_bit,
+            cycle_12_end_bit,
+
+            m,
+            t,
+            last_chunk_bit,
+            h_input,
+            h_output,
+            unused_row,
+
+            initial_hash,
+            initial_hash_compress,
+            inversion_const,
+            msg_chunks,
+            t_public,
+            last_chunk_bit_public,
+            hash_state,
+        }
+    }
+
+    fn cycle_12(&mut self) -> (BitRegister, BitRegister) {
+        let cycle_12_registers = self.alloc_array::<BitRegister>(CYCLE_12);
+
+        // Set the cycle 12 registers first row
+        self.set_to_expression_first_row(
+            &cycle_12_registers.get(0),
+            ArithmeticExpression::from_constant(L::Field::from_canonical_usize(1)),
+        );
+
+        for i in 1..CYCLE_12 {
+            self.set_to_expression_first_row(
+                &cycle_12_registers.get(i),
+                ArithmeticExpression::from_constant(L::Field::from_canonical_usize(0)),
+            );
+        }
+
+        // Set transition constraint for the cycle_12_registers
+        for i in 0..CYCLE_12 {
+            let next_i = (i + 1) % CYCLE_12;
+
+            self.set_to_expression_transition(
+                &cycle_12_registers.get(next_i).next(),
+                cycle_12_registers.get(i).expr(),
+            );
+        }
+
+        (
+            cycle_12_registers.get(0),
+            cycle_12_registers.get(CYCLE_12 - 1),
+        )
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    fn blake2b_compress(
+        &mut self,
+        m: &ArrayRegister<U64Register>,
+        h_input: &ArrayRegister<U64Register>,
+        h_output: &ArrayRegister<U64Register>,
+        iv_pub: &ArrayRegister<U64Register>,
+        inversion_const_pub: &U64Register,
+        t: &U64Register, // assumes t is not more than u64
```

Comment:
This is generally what we assume in function inputs

Reply:
Actually, I probably should of been more clear.  The blake2b specification does technically allow for message lengths of 2^128 (see here:  https://en.wikipedia.org/wiki/BLAKE_(hash_function)).

Also, for cases where the length is larger than 64 bits, it will xor the most sig 64 bits with an element in the work vector V.  And my implementation actually totally does away with that XOR b/c of the assumption.

I can add more clear comments about that.

Reply:
I see, yes we should probably document that in the code.

Reply:
Made a more clearer comment.